Software development tools for R

willbowditch opened the pull request:
Proposed talk for coffee and coding session - would be great to have some feedback on the content + any idea for anything I've missed. 

I'm not expecting this session to be a 'how to', but just give a general idea of the tools available and a few hints and tips for managing R projects.

willbowditch committed:
First draft of slides

willbowditch committed:
Change svg to png, some other edits

willbowditch committed:
Add generated slides

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+# Why use software development tools? 🤔
+
+- As analysts we care about getting the right answer, so the code needs to be correct and maintainable
+- Writing code can lead to lots of avoidable mistakes. There are a whole bunch of tools and practises from software development that can help us avoid mistakes
```
Writing code can lead to lots of avoidable mistakes. There are a whole bunch of tools and practices from software development that can help to prevent them.

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+  - (e.g. this is the version of the code used for release 2018)
+- **Who** made the changes and **why** is recorded
+- Working on the code at the same time is facilitated with 'branches' and 'merges'
+- If something breaks, `git` has a number of tools to find out where the break occurred
```
... and revert to previous versions if necessary

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+- Code can be restored to any point in time and can be tagged 
+  - (e.g. this is the version of the code used for release 2018)
+- **Who** made the changes and **why** is recorded
+- Working on the code at the same time is facilitated with 'branches' and 'merges'
```
I'm assuming you might expand on these when you're talking about them because they might be difficult to picture if you don't know git.

willbowditch replied:
Good point - I'm hoping the pictures will help clarify everything, I'll make sure I check for any questions at the end of the git section

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+
+
+
+- **Master** is a version of the code that is 'stable', the version that is used to produce the final output of a publication, for example. 
```
You then say features are worked on in a separate branch. So say that master is the 'stable' branch.

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+
+## Example `git` difference log
+
+How the code has changed is recorded at a ganular level. 
```
granular

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+
+![](./img/example_git_graph.png)
+
+Note that every change to the code, who made it, and when is recorded. Any of these snapshots of code could be tagged with a meaningful label, like 'this is the version of the code used for publication x'. 
```
Note that every change to the code includes, who made it, and when is recorded.

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+
+>In R, the fundamental unit of shareable code is the package. A package bundles together code, data, documentation, and tests, and is easy to share with others.
+
+I'm going to try and pursuade you that writing packages is worth the extra effort in the long run.
```
persuade

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
````diff
@@ -0,0 +1,454 @@
+**Q**: What happens if a statement fails part way through a script?
+
+```r
+mpg %>%
````
does everyone know what %>% means?

willbowditch replied:
Probably not... I'll chat it through in the session.

tamsinforbes committed:
Update README.md

Explicit function calls, line 294 - I don't think %>% will work unless it's package is loaded - but where did it come from???. Also I didn't know explicit function calls meant you didn't have to load the package - that's very useful - thanks!

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+
+
+
+## The good practise package
```
I think it's practice ?!?

K1Br commented on 20181010_Software_Dev_practices_in_R/README.md:
```diff
@@ -0,0 +1,454 @@
+
+## CI/CD
+
+- All of the above steps should be run on any code that is changed before it is merged into the 'master version'
```
Going back to Git...

tamsinforbes reviewed (commented):
made a couple of spelling corrections and commented on the %>% in explicit function calls section

willbowditch committed:
Changes suggested

willbowditch committed:
update schedule

willbowditch committed:
generate slides

willbowditch commented:
Thanks for the suggestions @tamsinforbes and @K1Br  👍🏻

I'll leave you to approve and merge, if that's alright @tamsinforbes

tamsinforbes reviewed (approved):
LGTM
